app/carpool/views.py

```diff
@@ -290,32 +299,53 @@ def _email_carpool_cancelled(carpool, reason, send=False):
 
     subject = 'Carpool session on {} cancelled'.format(carpool.leave_time)
 
-    # TODO: This should be an HTML template stored elsewhere
-    body = '''
-        Hello rider,
-
-        Unfortunately, the carpool session for leaving from {} at {} has been
-        cancelled.
-
-        The reason given for the cancellation was: {}.
-
-        Please reach out to {} in order to see if they're willing
-        to reschedule.
-    '''.format(
-            carpool.from_place,
-            carpool.leave_time,
-            reason,
-            driver.email)
+    messages_to_send = [
+        _make_email_message(
+            'carpools/email/carpool_cancelled.html',
+            'carpools/email/carpool_cancelled.txt',
+            rider.email,
+            subject,
+            driver=driver,
+            rider=rider,
+            carpool=carpool,
+            reason=reason) for rider in riders
+    ]
+
+    _send_email(messages_to_send, raise_connect_exceptions=True)
+
+
+def _make_email_message(html_template, text_template, recipient, subject,
+                        **kwargs):
+    body = render_template(text_template, **kwargs)
+    html = render_template(html_template, **kwargs)
+    return Message(
+        recipients=[recipient], body=body, html=html, subject=subject)
+
+
+def _send_email(messages_to_send, raise_connect_exceptions=False):
+    if current_app.config.get('MAIL_LOG_ONLY'):
+        current_app.logger.info(
+            'Configured to log {} messages without sending.  Messages in the following lines:'.
+            format(len(messages_to_send)))
+        for message in messages_to_send:
+            current_app.logger.info(
+                'Message to {} with subject {} and body {}'.format(
+                    message.recipients[0], message.subject, message.body))
+        return
 
-    if send:
+    try:
         with mail.connect() as conn:
-            for rider in riders:
-                msg = Message(recipients=[rider.email],
-                              body=body,
-                              subject=subject)
-                conn.send(msg)
-    else:
-        for rider in riders:
-            current_app.logger.info(
-                'Sent message to {} with subject {} and body {}'.format(
-                    rider.email, subject, body))
+            for message in messages_to_send:
+                try:
+                    conn.send(message)
+                except Exception as exception:
+                    current_app.logger.error(
+                        'Failed to send message to {} with subject {} and body {} Exception: {}'.
+                        format(message.recipients[0], message.subject,
+                               message.body, repr(exception)))
+    except Exception as exception:
+        if raise_connect_exceptions:
```

Comment:
I'd prefer to see the `_send_email()` function not do any exception handling so the caller can decide how what to do with the exceptions.

Reply:
By that I mean I'd like to see the `raise_connect_exceptions` arg removed and remove all the `try`/`except`s in here so that any exceptions get handled further up the call stack.

Reply:
@iandees if we handle exceptions further up the stack, that does mean that any single email-send exception means we can't send any subsequent emails...

Reply:
@iandees In most cases the code further up the stack doesn't care about email exceptions.  I did it this way to keep calling code uncluttered, to centralize logging and so we don't need to think about exception handling unless it's necessary.